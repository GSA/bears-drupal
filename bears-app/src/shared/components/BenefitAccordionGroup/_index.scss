#usagov-bears-app {
  .benefit-accordion-group {
    .benefit-accordion {
      margin-bottom: 26px;
      /* stylelint-disable selector-class-pattern */
      .usa-accordion__button {
        border-radius: 3px;
      }
      /* stylelint-enable selector-class-pattern */

      .benefit-detail-title {
        padding: 0;
        margin: 0 0 24px;
        text-align: left;
      }

      .benefit-detail-summary,
      .benefit-criteria-list,
      .benefit-alert {
        margin: 0 0 24px;
      }

      .unmet-criteria-group {
        margin: 0 0 24px;

        .unmet-criteria-title {
          width: 100%;
          text-align: center;
          color: #000;
          font-size: 16px;
          font-style: normal;
          font-weight: 400;
          line-height: 19.6px;
        }
      }

      .benefit-link {
        width: 100%;
        margin-bottom: 40px;
      }
    }

    .expand-all {
      font-size: 16px;
      font-weight: 400;
      margin-bottom: 32px;
    }
  }
}
