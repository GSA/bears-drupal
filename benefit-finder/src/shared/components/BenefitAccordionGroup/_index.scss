@use '../../styles/colors/_index.scss' as color;
@use '../../styles/mixins/_index.scss' as *;
@use '../../styles/space/_index.scss' as space;

#benefit-finder {
  .benefit-accordion-group {
    .benefit-accordion {
      margin-bottom: space.$space-md-plus;

      .usa-accordion__button {
        border-radius: 3px;
      }

      .benefit-detail-title {
        @include h6;

        padding: 0;
        margin: 0 0 24px;
        text-align: left;
      }

      .benefit-detail-summary,
      .benefit-criteria-list,
      .benefit-alert {
        @include p;

        margin: 0 0 24px;
      }

      .unmet-criteria-group {
        margin: 0 0 24px;

        .unmet-criteria-title {
          @include p;

          width: 100%;
          text-align: center;
          color: color.$pure-black;
        }
      }

      .benefit-link {
        width: 100%;
        margin-bottom: space.$space-lg;
      }
    }

    .expand-all {
      @include p;

      background-color: transparent !important;
      color: color.$pop-blue !important;
      margin-bottom: space.$space-lg;
    }
  }
}
