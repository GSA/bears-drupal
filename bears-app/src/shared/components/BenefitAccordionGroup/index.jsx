import { useState } from 'react'
import PropTypes from 'prop-types'
import {
  Accordion,
  Alert,
  Button,
  Heading,
  KeyElegibilityCrieriaList,
  ObfuscatedLink,
  Paragraph,
} from '../index'
import './_index.scss'

/**
 * a functional compound component that renders a benefit group accordion
 * @component
 * @param {array} data - our benefits data
 * @param {string} entryKey - which key in the array to target
 * @param {bool} expandAll - determnines if we include ExpandAll component
 * @return {html} returns html
 */
const BenefitAccordionGroup = ({ data, entryKey, expandAll }) => {
  /**
   * a hook that hanldes our open state of the accordions in our group
   * @function
   * @return {boolean} returns true or false
   */
  const [isExpandAll, setExpandAll] = useState(false)

  /**
   * a function that returns the string value of our expanded action
   * @function
   * @return {stroing} returns label for our button
   */
  const handleExpandIcon = isExpandAll ? 'Collapse all -' : 'Expand all +'

  /**
   * a functional component that renders a button and controls the expansion of our accordions
   * @component
   * @return {html} returns html
   */
  const ExpandAll = () => {
    return (
      expandAll && (
        <Button
          className="expand-all"
          aria-label={handleExpandIcon}
          unstyled
          onClick={() => setExpandAll(!isExpandAll)}
        >
          {handleExpandIcon}
        </Button>
      )
    )
  }

  /**
   * a functional component that list unmet criteria
   * @component
   * @param {array} item - an array of unmet criteria
   * @return {html} returns an unorderd list
   */
  const NotEligibleList = ({ items }) => {
    return (
      <div className="unmet-criteria-group">
        <div className="unmet-criteria-title">UNMET CRITERIA</div>
        <ul className="unmet-criteria-list">
          {items.map((item, index) => {
            const { label } = item
            return (
              <li key={index} className="unmet-criteria-item">
                {label}
              </li>
            )
          })}
        </ul>
      </div>
    )
  }

  /**
   * a functional component that list unmet criteria
   * @component
   * @param {array} item - an array of unmet criteria
   * @return {html} returns an unorderd list
   */
  const MoreInfoList = ({ items }) => {
    return (
      <div className="unmet-criteria-group">
        <div className="unmet-criteria-title">MORE INFORMATION NEEDED</div>
        <ul className="unmet-criteria-list">
          {items.map((item, index) => {
            const { label } = item
            return (
              <li key={index} className="unmet-criteria-item">
                {label}
              </li>
            )
          })}
        </ul>
      </div>
    )
  }

  return (
    <div className="benefit-accordion-group">
      <ExpandAll />
      {data &&
        data.map((item, index) => {
          const { agency, eligibility, sourceLink, summary, title } =
            item[entryKey]
          // filter to get benefit criteria matches
          const eligibleBenefits = eligibility.filter(
            item => item.isEligible === true
          )
          // filter to get unmet criteria
          const notEligibleBenefits = eligibility.filter(
            item => item.isEligible === false || item.isEligible === undefined
          )

          // filter to get criteria without user values to compare with
          const moreInformationNeeded = eligibility.filter(
            item => item.isEligible === 'undefined'
          )
          // determine the eligibility statues based on the benefits length
          // Total Criteria = y
          // Met Criteria = x
          // Not Met Criteria = z

          // Criteria Met Length	Label
          // x === y	"Likely Eligible"
          // x !== 0 && x < y	"Potentially Eligible"
          // x === undefined "More Information Needed"
          // Criteria Not Met	Label
          // z > 0	"Not Eligible"
          const eligibleStatus =
            eligibleBenefits.length === eligibility.length
              ? 'Likely Eligible'
              : eligibleBenefits.length > 0 &&
                eligibleBenefits.length < eligibility.length &&
                moreInformationNeeded.length === 0
              ? 'Potentially  Eligible'
              : moreInformationNeeded.length > 0
              ? 'More Information Needed'
              : 'Not Eligible'

          return (
            <Accordion
              key={`${index}-${title}`}
              id={`${index}-${title}`}
              heading={title}
              subHeading={eligibleStatus}
              aria-expanded={isExpandAll}
              isExpanded={isExpandAll}
              data-analytics="benefit-accordion"
              data-analytics-content={title}
            >
              <Heading className="benefit-detail-title" headingLevel={4}>
                {`Provided by ${agency.title}`}
              </Heading>
              <Paragraph className="benefit-detail-summary">
                {summary}
              </Paragraph>
              <KeyElegibilityCrieriaList
                className="benefit-criteria-list"
                data={eligibleBenefits}
                initialEligibilityLength={eligibility.length}
              />
              {notEligibleBenefits.length > 0 && (
                <NotEligibleList items={notEligibleBenefits} />
              )}
              {moreInformationNeeded.length > 0 && (
                <MoreInfoList items={moreInformationNeeded} />
              )}
              <Alert className="benefit-alert">
                Additional eligibility criteria may apply. Please visit agency
                for full requirements.
              </Alert>
              <ObfuscatedLink className="benefit-link" href={sourceLink}>
                Visit {agency.title}
              </ObfuscatedLink>
            </Accordion>
          )
        })}
    </div>
  )
}

BenefitAccordionGroup.propTypes = {
  data: PropTypes.array,
  entryKey: PropTypes.string,
  expandAll: PropTypes.bool,
}

export default BenefitAccordionGroup
