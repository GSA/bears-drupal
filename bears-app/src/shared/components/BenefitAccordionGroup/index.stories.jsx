import BenefitAccordionGroup from './index.jsx'
import * as DATA from '../../api/mock-data/content-data.json'

const data = [DATA.benefits[0], DATA.benefits[0]]
const entryKey = Object.keys(data[0])

export default {
  component: BenefitAccordionGroup,
  tags: ['autodocs'],
  parameters: {
    layout: 'fullscreen',
  },
  args: {
    data,
    entryKey: entryKey[0],
  },
}

export const Primary = {}

export const ExpandAll = {
  args: {
    ...Primary.args,
    expandAll: true,
  },
}
