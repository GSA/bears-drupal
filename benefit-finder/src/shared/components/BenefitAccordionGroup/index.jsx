import { useState } from 'react'
import PropTypes from 'prop-types'
import createMarkup from '../../utils/createMarkup'
import {
  Accordion,
  Alert,
  Button,
  Heading,
  KeyElegibilityCrieriaList,
  ObfuscatedLink,
} from '../index'
import './_index.scss'

/**
 * a functional compound component that renders a benefit group accordion
 * @component
 * @param {array} data - our benefits data
 * @param {string} entryKey - which key in the array to target
 * @param {bool} expandAll - determnines if we include ExpandAll component
 * @return {html} returns html
 */
const BenefitAccordionGroup = ({
  data,
  entryKey,
  expandAll,
  notQualifiedView,
  ui,
}) => {
  const { benefitAccordion, benefitAccordionGroup } = ui
  const {
    eligibleStatusLabels,
    agencyPrefix,
    visitLabel,
    unmetLabel,
    additionalDescription,
  } = benefitAccordion
  const { closedState, openState } = benefitAccordionGroup

  // debugger
  if (
    window.location.hostname === 'bears-waf.app.cloud.gov' ||
    window.location.hostname === 'localhost'
  ) {
    console.log(data && data.length > 0 && { benefitEligibility: data })
    // data &&
    //   data.forEach(item => {
    //     console.log('---')
    //     console.log(item.benefit.title)
    //     console.table(item.benefit.eligibility)
    //     console.log('---')
    //   })
  }
  /**
   * a hook that hanldes our open state of the accordions in our group
   * @function
   * @return {boolean} returns true or false
   */
  const [isExpandAll, setExpandAll] = useState(false)
  // const [filterData, setFilterData] = useState(null)

  /**
   * a function that returns the string value of our expanded action
   * @function
   * @return {stroing} returns label for our button
   */
  const handleExpandIcon = isExpandAll ? `${openState} -` : `${closedState} +`

  /**
   * a functional component that renders a button and controls the expansion of our accordions
   * @component
   * @return {html} returns html
   */
  const ExpandAll = () => {
    return (
      expandAll && (
        <Button
          className="expand-all"
          aria-label={handleExpandIcon}
          unstyled
          onClick={() => setExpandAll(!isExpandAll)}
        >
          {handleExpandIcon}
        </Button>
      )
    )
  }

  /**
   * a functional component that list unmet criteria
   * @component
   * @param {array} item - an array of unmet criteria
   * @return {html} returns an unorderd list
   */
  const NotEligibleList = ({ items }) => {
    return (
      <div className="unmet-criteria-group">
        <div className="unmet-criteria-title">{unmetLabel}</div>
        <ul className="unmet-criteria-list">
          {items.map((item, index) => {
            const { label } = item
            return (
              <li
                key={`not-eligible-list-${index}`}
                className="unmet-criteria-item"
              >
                {label}
              </li>
            )
          })}
        </ul>
      </div>
    )
  }

  /**
   * a functional component that list unmet criteria
   * @component
   * @param {array} item - an array of unmet criteria
   * @return {html} returns an unorderd list
   */
  const MoreInfoList = ({ items }) => {
    return (
      <div className="unmet-criteria-group">
        <div className="unmet-criteria-title">{eligibleStatusLabels[1]}</div>
        <ul className="unmet-criteria-list">
          {items.map((item, index) => {
            const { label } = item
            return (
              <li key={`more-info-${index}`} className="unmet-criteria-item">
                {label}
              </li>
            )
          })}
        </ul>
      </div>
    )
  }

  return (
    <div className="benefit-accordion-group">
      <ExpandAll />
      {data &&
        data.map((item, index) => {
          const { agency, eligibility, SourceLink, summary, title } =
            item[entryKey]
          // filter to get benefit criteria matches
          const eligibleBenefits = eligibility.filter(
            item => item.isEligible === true
          )
          // filter to get unmet criteria
          const notEligibleBenefits = eligibility.filter(
            item => item.isEligible === false
          )

          // filter to get criteria without user values to compare with
          const moreInformationNeeded = eligibility.filter(
            item => item.isEligible === undefined
          )
          // determine the eligibility statues based on the benefits length
          // Total Criteria = y
          // Met Criteria = x
          // Not Met Criteria = z

          // Criteria Met Length	Label
          // x === y	"Likely Eligible"
          // z === 0 && x === undefined length > 0 "More Information Needed"
          // Criteria Not Met	Label
          // z > 0	"Not Eligible"
          const eligibleStatus =
            eligibleBenefits.length === eligibility.length
              ? eligibleStatusLabels[0]
              : notEligibleBenefits.length === 0 &&
                moreInformationNeeded.length > 0
              ? eligibleStatusLabels[1]
              : eligibleStatusLabels[2]

          const handleHidden =
            notQualifiedView === false &&
            eligibleStatus !== eligibleStatusLabels[0]
              ? true
              : !!(
                  notQualifiedView === true &&
                  eligibleStatus === eligibleStatusLabels[0]
                )

          return (
            <Accordion
              key={`${index}-${title}`}
              id={`${index}-${title}`}
              heading={title}
              subHeading={eligibleStatus}
              aria-expanded={isExpandAll}
              isExpanded={isExpandAll}
              data-analytics="benefit-accordion"
              data-analytics-content={title}
              hidden={handleHidden}
              data-testid="benefit"
            >
              <Heading className="benefit-detail-title" headingLevel={4}>
                {`${agencyPrefix} ${agency.title}`}
              </Heading>
              <div
                className="benefit-detail-summary"
                dangerouslySetInnerHTML={createMarkup(summary)}
              />
              <KeyElegibilityCrieriaList
                className="benefit-criteria-list"
                data={eligibleBenefits}
                initialEligibilityLength={eligibility.length}
                ui={benefitAccordion}
              />
              {notEligibleBenefits.length > 0 && (
                <NotEligibleList items={notEligibleBenefits} />
              )}
              {moreInformationNeeded.length > 0 && (
                <MoreInfoList items={moreInformationNeeded} />
              )}
              <Alert className="benefit-alert">{additionalDescription}</Alert>
              <ObfuscatedLink
                className="benefit-link"
                href={SourceLink}
                target="_blank"
                rel="noopener noreferrer"
              >
                {visitLabel} {agency.title}
              </ObfuscatedLink>
            </Accordion>
          )
        })}
    </div>
  )
}

BenefitAccordionGroup.propTypes = {
  data: PropTypes.array,
  entryKey: PropTypes.string,
  expandAll: PropTypes.bool,
}

export default BenefitAccordionGroup
