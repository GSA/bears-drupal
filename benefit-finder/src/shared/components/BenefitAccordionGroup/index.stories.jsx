import BenefitAccordionGroup from './index.jsx'
import content from '../../api/mock-data/current.js'

const { data } = JSON.parse(content)
const { benefits } = data
const b = [benefits[0], benefits[1]]
const entryKey = Object.keys(b[0])

export default {
  component: BenefitAccordionGroup,
  tags: ['autodocs'],
  args: {
    data: b,
    entryKey: entryKey[0],
  },
}

export const Primary = {}

export const ExpandAll = {
  args: {
    ...Primary.args,
    expandAll: true,
  },
}
