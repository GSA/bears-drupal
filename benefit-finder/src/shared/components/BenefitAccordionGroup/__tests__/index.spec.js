import renderer from 'react-test-renderer'
import BenefitAccordionGroup from '../index.jsx'

describe('BenefitAccordionGroup', () => {
  it('renders a match to the previous snapshot', () => {
    const component = renderer.create(<BenefitAccordionGroup />)
    expect(component.toJSON()).toMatchSnapshot()
  })
})
