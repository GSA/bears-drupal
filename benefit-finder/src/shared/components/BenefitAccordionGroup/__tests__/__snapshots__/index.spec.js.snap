// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BenefitAccordionGroup renders a match to the previous snapshot 1`] = `
<div
  className="benefit-accordion-group"
/>
`;
